src/main/java/org/jenkinsci/plugins/ssegateway/EventHistoryStore.java

```diff
@@ -0,0 +1,297 @@
+        if (autoExpireTimer != null) {
+            return;
+        }
+        
+        // Set up a timer that runs the DeleteStaleHistoryTask 3 times over the
+        // duration of the event expiration timeout. By default this will be
+        // every 20 seconds i.e. in that case, events are never left lying around
+        // for more than 20 seconds past their expiration.
+        long taskSchedule = expiresAfter / 3;
+        autoExpireTimer = new Timer();
+        autoExpireTimer.schedule(new DeleteStaleHistoryTask(), taskSchedule, taskSchedule);
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                disableAutoDeleteOnExpire();
+            }
+        });
+    }
+    
+    public synchronized static void disableAutoDeleteOnExpire() {
+        if (autoExpireTimer == null) {
+            return;
+        }
+        autoExpireTimer.cancel();
+        autoExpireTimer = null;
+    }
+    
+    private static class DeleteStaleHistoryTask extends TimerTask {
+        @Override
+        public void run() {
+            try {
+                deleteStaleHistory();
+            } catch (Exception e) {
+                LOGGER.log(Level.SEVERE, "Error deleting stale/expired events from EventHistoryStore.", e);
+            }
+        }
+    }
+    
+    private static synchronized AtomicInteger getChannelSubsCounter(@Nonnull String channelName) {
+        AtomicInteger counter = channelSubsCounters.get(channelName);
+        if (counter == null) {
+            counter = new AtomicInteger(0);
+            channelSubsCounters.put(channelName, counter);
+        }
+        return counter;
+    }
+
+    /**
+     * This class is used to listen for and log/store events that are being deivered
+     * on any channels for which there are subscriptions ({@link EventDispatcher}).
+     * We do this here because we do not want every {@link EventDispatcher} taking care
+     * of the storing of events that might need to be retried. Instead, we have one
+     * listener that does that (an instance of this class) and then all the
+     * {@link EventDispatcher} listeners just hold a retryQueue that contains references
+     * to the event UUIDs and use that to access the {@link EventHistoryStore} to get
+     * the actual message when doing the retry. So, one listener storing the events for
+     * all the {@link EventDispatcher} instances.
+     * <p>
+     * Note, we never remove these listeners even if the channel in question no
+     * longer has any active subscribers. There's no real point.
+     */
+    private static class EventHistoryLogger implements ChannelSubscriber {
+
+        private final AtomicInteger channelSubsCounter;
+
+        private EventHistoryLogger(AtomicInteger channelSubsCounter) {
+            this.channelSubsCounter = channelSubsCounter;
+        }
+
+        @Override
+        public void onMessage(@Nonnull Message message) {
```

Comment:
No, it's called async from the messaging layer and each call to the store will store a separate file based on the UUID of the message i.e. guaranteed that only one thread is ever writing to the file at that time.

That said ... there actually might be a tiny window where the thread writing the file (of which there can only be one) may have created the file but not be finished writing it, while at the same time the retry queue thread tries to read that message. I might look at doing something more atomic to try prevent that. Thanks man !!